kbs/docs/initdata.md

```diff
@@ -0,0 +1,237 @@
+the TEE when launching. We do not limit the encoding of this data structure, which
+means that JSON, TOML and YAML are optional. This will be introduced in [Initdata](#initdata)
+- A data integrity binding mechanism. It will guide the untrusted host to bind the
+digest of the `data` part (initdata data) in the Initdata to the hardware TEE-specific 
+field in evidence. This field will be checked by the verifier during the remote
+attestation. This will be introduced in 
+[Integrity Binding for Different TEEs](#integrity-binding-for-different-tees)
+- A data serialization method. This method is used to serialize the Initdata data
+into a canonicalized one, which will help to get consistent cryptographic hash for
+different forms of a same initdata. This will be introduced in
+[Data Canonicalization Algorithm](#data-canonicalization-algorithm).
+
+This spec does not define how the initdata will be delivered into the TEE.
+Different projects will have its own way to do this. For Confidential Containers,
+we will use kata-runtime and kata-agent to collaborate to achieve this function.
+
+## Terminology
+
+This section will introduce the terminology used in this spec to avoid ambiguity.
+
+- `Initdata`: A data structure that includes initdata data and other information
+that will help to calculate the initdata digest. The whole data structure will
+be delivered into the guest.
+- `Initdata Metadata`: Metadata fields of initdata, s.t. `algorithm`, `version`,
+etc. They are used to calculate the initdata digest.
+- `Initdata data`: Data that needs to be injected when the TEE is started. This data
+requires integrity protection but does not require confidentiality protection. In
+initdata, this will be included inside the `data` section.
+- `Initdata digest`: Digest of the initdata data calculated following this spec.
+It will be used as the value of the TEE hostdata/initdata field.
+- `TEE initdata/hostdata`: Fields that can be bound to a specific TEE instance. This field
+information will be included in the TEE-signed remote attestation report. Typically,
+Intel TDX's `mr_config_id`, AMD SNP's `hostdata` and Arm CCA's `CCA_REALM_PERSONALIZATION_VALUE`.
+In order to avoid confusion with the hostdata field of AMD SNP, when we do not
+emphasize a specific platform s.t. SNP, we are referring to the corresponding fields of
+various TEE platforms.
+
+## Specifications
+
+### Initdata
+
+Initdata defines a standardized structure format. Please note that it
+does not indicate the specific encoding format, but requires that the encoding format
+must support the expression of key-value data pairs. Typical encodings that meet
+this requirement include JSON, TOML and YAML, etc.
+
+An initdata SHOULD have the following fields
+- `version`: The format version of the initdata metadata. Version number will provide
+extensibility. The definition in this spec is all `0.1.0`.
+- `algorithm`: The hash algorithm to calculate the value to set as `HOSTDATA`. The typical
```

Comment:
Good point. How to fill into the hw initdata field is a problem. To resolve the same problem, another way is to define truncation and padding rules.

That means if we specify sha512 and get a 64 bytes data. We will truncate to 32 bytes to set a SNP hostdata (32 bytes length).
Or, if we specify sha256 and get a 32 bytes data. We will padding b"0" to 64 bytes to set a Arm CCA `CCA_REALM_PERSONALIZATION_VALUE` (64 bytes length)

When comparing against the expect value, applying the same rule.

- The advantage is that we would not need to care the underlying TEE type when selecting hash algorithms. This means the TEE type can be transparent to users. Also, we do not bind a specific hash algorithm to a specific platform, which can be flexible.
- The disadvantage is that truncating the hash will reduce the security of the original hash function according to the number of truncated bits (increasing the possibility of collision), while it is [ok](https://crypto.stackexchange.com/questions/9435/is-truncating-a-sha512-hash-to-the-first-160-bits-as-secure-as-using-sha1).

Reply:
Interesting. I guess the question is how we reconcile the hash algorithm with the platform. In theory we might want to use the same initdata for several different platforms. There are a couple of things we could do.

1) Have an algorithm field for each platform. This makes the calculation of the hash a bit more complicated but generally seems sound.

2) Only use one hash algorithm field but allow padding and truncation. This again adds some complexity but not much platform-specific logic. 

3) Dynamic algorithm field. We could add logic to the kata shim to select the algorithm on the fly based on the platform. We'd have to think through how this would affect reference values.

I think I lean towards 2 atm, but I need to think about this more.

Reply:
Ok, another thing that's nice about the second option is that it doesn't require the client to know much about the platforms that the initdata might be used with. With option 1 it seems like they would have to come up with the various algs for every platform that they might want to use. Granted that's not too hard especially if there is a tool to generate it, but it seems a little nicer to just have one alg field and then take care of fitting it into the host data automatically.

Reply:
Let me double check: the conclusion is we only have one hash algorithm and define padding and truncation.

Reply:
That's my preference. @danmihai1?

Reply:
@Xynnn007 @fitzthum I think it's a shame to have more bits available and to not use them. For example, if sha256 would have been enough for solutions with high security/confidentiality requirements, I don't think cryptographers would have invented sha384. So, I think we should use all the bits available for a given TEE type and the most secure digest algorithm corresponding to that bit length.

Reply:
@danmihai1 For this, we can use a `sha512` by default. The truncation of sha value is secure as mentioned in [this](https://crypto.stackexchange.com/questions/9435/is-truncating-a-sha512-hash-to-the-first-160-bits-as-secure-as-using-sha1). That is to say, the first 384 bits of sha512 is the same secure as sha384. The good property would allow us not to define different hash algorithms.

Reply:
@Xynnn007 thanks for that reminder!

I would split this discussion in two separate discussions:

a. The format of the Init Data file
b. Handling of the TEE field by the CoCo Host and Guest software

For (a), I would remove the digest field from the file specification, because the CoCo Host can compute that value and the Guest MUST compute it. So, it doesn't seem like this field is required.

For (b): 

- The behavior of digest algorithm(s) would be specified by the relevant software components from the CoCo Host and Guest. For example, the Kata Shim and the Kata Agent would specify the digest algorithm(s) they support. The trusted Kata Agent and remote attestation server must agree on the digest algorithm(s). So, these two components also dictate the digest algorithm(s) to the Kata Shim.

thanks for that reminder!

I would split this discussion in two separate discussions:

a. The format of the Init Data file
b. The TEE field handling by the CoCo Host and Guest

For (a), I would remove the digest field from the file specification, because the CoCo Host can compute that value and the Guest MUST compute it. So, it doesn't seem like this field is required.

For (b): 

- The behavior of digest algorithm(s) would be specified by the relevant software components from the CoCo Host and Guest. For example, the Kata Shim and the Kata Agent would specify the digest algorithm(s) they support. The trusted Kata Agent and remote attestation server must agree on the digest algorithm(s). So, these two components also dictate the digest algorithm(s) to the Kata Shim.

- I would use use SHA-256 for SNP Host Data and SHA-384 for TDX MRCONFIGID, because these algorithms seem to be widely accepted - e.g., for FIPS 140 certification, etc. I don't know how complex and widely accepted digest truncation algorithms are.

- I don't see many advantages in forcing SNP and TDX to use the same algorithm, especially if the truncation will work differently across those two platforms.

- From Microsoft's point of view, SHA-256 for SNP Host Data would work best, because it would be consistent with already existing Microsoft products.

Reply:
@fitzthum @Xynnn007 I've talked with my colleagues Ken Gordon, Bryce Fisher, and others about the model described by Tobin (SHA256 + padding/reserved bits). Ken pointed out that model helps with using a consistent key release policy across SNP and TDX. So, we concluded that would be a good model - thanks Tobin!

(That conclusion goes against some of what I said above. So, it looks like I should talk with my colleagues - smarter than me! - more often.)